fabric-api-base/src/testmodClient/java/net/fabricmc/fabric/test/base/client/FabricClientTestHelper.java

```diff
@@ -182,41 +184,43 @@ public static void waitForTitleScreenFade() {
 		});
 	}
 
-	private static void waitFor(String what, Predicate<MinecraftClient> predicate) {
-		waitFor(what, predicate, Duration.ofSeconds(10));
+	public static void waitForServerStop() {
+		waitFor("Server stop", client -> !ThreadingImpl.isServerRunning, SharedConstants.TICKS_PER_MINUTE);
 	}
 
-	private static void waitFor(String what, Predicate<MinecraftClient> predicate, Duration timeout) {
-		final LocalDateTime end = LocalDateTime.now().plus(timeout);
-
-		while (true) {
-			boolean result = submitAndWait(predicate::test);
+	private static void waitFor(String what, Predicate<MinecraftClient> predicate) {
+		waitFor(what, predicate, 10 * SharedConstants.TICKS_PER_SECOND);
+	}
 
-			if (result) {
-				break;
-			}
+	private static void waitFor(String what, Predicate<MinecraftClient> predicate, int timeoutTicks) {
+		int tickCount;
 
-			if (LocalDateTime.now().isAfter(end)) {
-				throw new RuntimeException("Timed out waiting for " + what);
-			}
+		for (tickCount = 0; tickCount < timeoutTicks && !computeOnClient(predicate::test); tickCount++) {
+			runTick();
+		}
 
-			waitFor(Duration.ofMillis(50));
+		if (tickCount == timeoutTicks && !computeOnClient(predicate::test)) {
+			throw new RuntimeException("Timed out waiting for " + what);
 		}
 	}
 
-	private static void waitFor(Duration duration) {
-		try {
-			Thread.sleep(duration.toMillis());
-		} catch (InterruptedException e) {
-			throw new RuntimeException(e);
+	public static void runTicks(int ticks) {
+		for (int i = 0; i < ticks; i++) {
+			runTick();
 		}
 	}
 
-	private static <T> CompletableFuture<T> submit(Function<MinecraftClient, T> function) {
-		return MinecraftClient.getInstance().submit(() -> function.apply(MinecraftClient.getInstance()));
+	public static void runTick() {
+		ThreadingImpl.runTick();
+	}
+
+	public static <E extends Throwable> void runOnClient(FailableConsumer<MinecraftClient, E> action) throws E {
+		ThreadingImpl.runOnClient(() -> action.accept(MinecraftClient.getInstance()));
 	}
 
-	public static <T> T submitAndWait(Function<MinecraftClient, T> function) {
-		return submit(function).join();
+	public static <T, E extends Throwable> T computeOnClient(FailableFunction<MinecraftClient, T, E> action) throws E {
+		MutableObject<T> result = new MutableObject<>();
+		runOnClient(minecraft -> result.setValue(action.apply(minecraft)));
+		return result.getValue();
 	}
```

Comment:
Suggestion: Calling it `getFromClient()` instead of `computeOnClient()` would make it clearer that this method returns a value while `runOnClient()` doesn't.

Your use of `compute` here isn't wrong, I believe the Java API does something similar, but I think many people use `run` and `compute` synonymously and wouldn't immediately grasp that this is the difference between them. `get` vs `run` makes it really obvious.

Reply:
I'm pretty sure this is the more common terminology, and it should be immediately obvious what the difference between the two functions are from the parameter type

Reply:
I do like simple language, but yeah `compute` is pretty common too. I don't think parameter/return types are usually the first thing I would look at when working with a new API, but then again this will eventually have Javadoc that can make it obvious more quickly.